drivers/firmware/ti_sci/ti_sci.c

```diff
@@ -0,0 +1,1660 @@
+	struct ti_sci_desc desc;
+	struct ti_sci_version_info version;
+	uint32_t host_id;
+	uint8_t seq;
+	struct k_sem data_sem;
+};
+
+/* Core/Setup Functions */
+static struct ti_sci_xfer *ti_sci_setup_one_xfer(const struct device *dev, uint16_t msg_type,
+						 uint32_t msg_flags, void *req_buf,
+						 size_t tx_message_size, void *resp_buf,
+						 size_t rx_message_size)
+{
+	struct tisci_data *data = dev->data;
+	struct ti_sci_xfer *xfer = &data->xfer;
+	struct ti_sci_msg_hdr *hdr;
+
+	if (rx_message_size > data->desc.max_msg_size ||
+	    tx_message_size > data->desc.max_msg_size ||
+	    (rx_message_size > 0 && rx_message_size < sizeof(*hdr)) ||
+	    tx_message_size < sizeof(*hdr)) {
+		return NULL;
+	}
+
+	k_sem_take(&data->data_sem, K_FOREVER);
+	data->seq++;
+	k_sem_give(&data->data_sem);
+
+	xfer->tx_message.data = req_buf;
+	xfer->tx_message.size = tx_message_size;
+	xfer->rx_message.buf = resp_buf;
+	xfer->rx_message.size = (uint8_t)rx_message_size;
+
+	hdr = (struct ti_sci_msg_hdr *)req_buf;
+	hdr->seq = data->seq;
+	hdr->type = msg_type;
+	hdr->host = data->host_id;
+	hdr->flags = msg_flags;
+
+	return xfer;
+}
+
+static void callback(const struct device *dev, mbox_channel_id_t channel_id, void *user_data,
+		     struct mbox_msg *data)
+{
+	k_sem_give(&response_ready_sem);
+}
+
+static bool ti_sci_is_response_ack(void *r)
+{
+	struct ti_sci_msg_hdr *hdr = (struct ti_sci_msg_hdr *)r;
+
+	return hdr->flags & TI_SCI_FLAG_RESP_GENERIC_ACK ? true : false;
+}
+
+static int ti_sci_get_response(const struct device *dev, struct ti_sci_xfer *xfer)
+{
+	struct tisci_data *dev_data = dev->data;
+	struct ti_sci_msg_hdr *hdr;
+	int ret = 0;
+
+	if (!xfer->rx_message.buf) {
+		LOG_ERR("No response buffer provided");
+		return -EINVAL;
+	}
+
+	if (k_sem_take(&response_ready_sem, K_MSEC(dev_data->desc.max_rx_timeout_ms * 1000)) != 0) {
+		LOG_ERR("Timeout waiting for response");
+		return -ETIMEDOUT;
+	}
+
+	hdr = (struct ti_sci_msg_hdr *)rx_message.buf;
+
+	/* Sanity check for message response */
+	if (hdr->seq != dev_data->seq) {
+		LOG_ERR("HDR seq != data seq [%d != %d]\n", hdr->seq, dev_data->seq);
+		return -EINVAL;
+	}
+
+	if (rx_message.size > dev_data->desc.max_msg_size) {
+		LOG_ERR("rx_message.size [ %d ] > max_msg_size\n", xfer->rx_message.size);
+		return -EINVAL;
+	}
+
+	if (rx_message.size < xfer->rx_message.size) {
+		LOG_ERR("rx_message.size [ %d ] < xfer->rx_message.size\n", xfer->rx_message.size);
+		return -EINVAL;
+	}
+
+	memcpy(xfer->rx_message.buf, rx_message.buf, xfer->rx_message.size);
+
+	return ret;
+}
+
+static int ti_sci_do_xfer(const struct device *dev, struct ti_sci_xfer *xfer)
+{
+	const struct tisci_config *config = dev->config;
+	struct mbox_msg *msg = &xfer->tx_message;
+	int ret;
+
+	ret = mbox_send_dt(&config->mbox_tx, msg);
+	if (ret < 0) {
+		LOG_ERR("Could not send (%d)\n", ret);
+		return ret;
+	}
+
+	/* Get response if requested */
+	if (xfer->rx_message.size) {
+		return ti_sci_get_response(dev, xfer);
```

Comment:
Are you sure you want to always `return` here? Seems this would cause the `ti_sci_is_response_ack()` check below to be skipped in cases where we return data (which needs the ACK check also). Maybe moving the `ti_sci_is_response_ack()` above this to line 164 would be the right solution.